Code/GraphMol/CIPLabeler/Digraph.cpp

```diff
@@ -0,0 +1,205 @@
+  visit[atom->getIdx()] = 1;
+
+  auto dist = 1;
+  auto flags = 0x0;
+  auto atomic_num = atom->getAtomicNum();
+
+  dp_root = &addNode(std::move(visit), atom, atomic_num, dist, flags);
+  dp_origin = dp_root;
+}
+
+const CIPMol &Digraph::getMol() const { return d_mol; };
+
+Node *Digraph::getOriginRoot() const { return dp_origin; };
+
+Node *Digraph::getCurrentRoot() const { return dp_root; }
+
+int Digraph::getNumNodes() const { return d_nodes.size(); }
+
+std::vector<Node *> Digraph::getNodes(Atom *atom) const {
+  std::vector<Node *> result;
+  auto queue = std::list<Node *>({getCurrentRoot()});
+
+  for (auto itr = queue.begin(); itr != queue.end(); ++itr) {
+    const auto &node = *itr;
+
+    if (atom == node->getAtom()) {
+      result.push_back(node);
+    }
+    for (const auto &e : node->getEdges()) {
+      if (!e->isBeg(node)) {
+        continue;
+      }
+      queue.push_back(e->getEnd());
+    }
+  }
+  return result;
+}
+
+/**
+ * Access the reference atom for Rule 6 (if one is set).
+ */
+Atom *Digraph::getRule6Ref() const { return dp_rule6Ref; }
+
+/**
+ * Used exclusively for Rule 6, we set one atom as the reference.
+ * @param ref reference atom
+ */
+void Digraph::setRule6Ref(Atom *ref) { dp_rule6Ref = ref; }
+
+/**
+ * Sets the root node of this digraph by flipping the directions
+ * of edges as required.
+ *
+ * @param newroot the new root
+ */
+void Digraph::changeRoot(Node *newroot) {
+
+  std::vector<Edge *> toflip;
+  auto queue = std::list<Node *>({newroot});
+  for (auto itr = queue.begin(); itr != queue.end(); ++itr) {
+    const auto &node = *itr;
+
+    for (const auto &e : node->getEdges()) {
+      if (e->isEnd(node)) {
+        toflip.push_back(e);
+        queue.push_back(e->getBeg());
+      }
+    }
+  }
+  for (auto &e : toflip) {
+    e->flip();
+  }
+  dp_root = newroot;
+}
+
+void Digraph::expand(Node *beg) {
+  const auto &atom = beg->getAtom();
+  const auto &edges = beg->getEdges();
+  const auto &prev =
+      edges.size() > 0 && !edges[0]->isBeg(beg) ? edges[0]->getBond() : nullptr;
+
+  if (MAX_NODE_DIST > 0 && beg->getDistance() > MAX_NODE_DIST) {
+    return;
+  }
+  if (MAX_NODE_COUNT > 0 && d_nodes.size() >= MAX_NODE_COUNT) {
+    std::stringstream errmsg;
+    errmsg << "Digraph generation failed: more than " << MAX_NODE_COUNT
+           << "nodes found.";
+    throw TooManyNodesException(errmsg.str());
+  }
+
+  // create 'explicit' nodes
+  for (const auto &bond : d_mol.getBonds(atom)) {
+    const auto &nbr = bond->getOtherAtom(atom);
+    const int nbrIdx = nbr->getIdx();
+    const int bord = d_mol.getBondOrder(bond);
+    const int virtual_nodes = bord - 1;
+
+    if (!beg->isVisited(nbrIdx)) {
+      auto end = beg->newChild(nbrIdx, nbr);
+      addEdge(beg, bond, end);
+
+      // duplicate nodes for bond orders (except for root atoms...)
+      // for example >S=O
+      if (dp_origin != beg) {
+        if (atom->getFormalCharge() < 0 &&
```

Comment:
Why is a negative formal charge significant here?

Reply:
I'm not sure, it is directly correspondent with the Java code, but given that the other part of the conditional clause requires it to have a fractional atomic number, I'd say it has something to do with consistently resolving mancude rings.